In [1]:
# today's agenda: the ITIS database can be accessed/loaded locally

import sqlite3
import pandas as pd

csv_file = 'comptab_2018-01-29 16_00_comma_separated.csv'
df_csv = pd.read_csv(csv_file)
df_csv.tail()

Year App.                Taxon     Class          Order       Family  \
67156  2017  III         Canis aureus  Mammalia      Carnivora      Canidae   
67157  2017  III    Sarcoramphus papa      Aves  Falconiformes  Cathartidae   
67158  2017  III      Daboia russelii  Reptilia      Serpentes    Viperidae   
67159  2017  III  Civettictis civetta  Mammalia      Carnivora   Viverridae   
67160  2017  III  Civettictis civetta  Mammalia      Carnivora   Viverridae   

              Genus Importer Exporter Origin  Importer reported quantity  \
67156         Canis       US       ET    NaN                         NaN   
67157  Sarcoramphus       JP       PE    NaN                         NaN   
67158        Daboia       RS       IT     US                       200.0   
67159   Civettictis       FR       ET    NaN                         NaN   
67160   Civettictis       KR       ET    NaN                         NaN   

       Exporter reported quantity       Term Unit Purpose Source  
67156                         7.0   trophies  NaN       H      W  
67157                         2.0       live  NaN       T      F  
67158                         NaN  specimens  NaN       T      F  
67159                       635.1       musk   kg       T      R  
67160                       480.0       musk   kg       T      R

In [2]:
print(df_csv['Term'].unique())

['bodies' 'feathers' 'specimens' 'live' 'caviar' 'skins' 'seeds' 'hair'
 'carvings' 'baleen' 'bone pieces' 'bone carvings' 'teeth' 'skin pieces'
 'meat' 'bones' 'jewellery' 'skeletons' 'trophies' 'horns' 'skulls'
 'fur product (small)' 'garments' 'dried plants' 'extract' 'cloth'
 'unspecified' 'shells' 'eggshell' 'carapaces' 'leather products (small)'
 'oil' 'eggs' 'scales' 'genitalia' 'medicine' 'roots' 'powder'
 'derivatives' 'leather products (large)' 'rug' 'tails' 'cosmetics'
 'ivory carvings' 'ivory pieces' 'piano keys' 'tusks' 'jewellery - ivory '
 'feet' 'ears' 'claws' 'eggs (live)' 'wood product' 'sawn wood' 'veneer'
 'musk' 'cultures' 'stems' 'flowers' 'fins' 'horn carvings' 'horn pieces'
 'swim bladders' 'gall' 'coral sand' 'fingerlings' 'raw corals' 'sides'
 'plates' 'leaves' 'fibres' 'timber' 'hair products'
 'fur products (large)' 'chips' 'trunk' 'wax' 'logs' 'pupae'
 'gall bladders' 'bark' 'fruit' 'pearl']


In [3]:
db_path = 'ITIS.sqlite'
conn = sqlite3.connect(db_path)

query = "SELECT tsn, completename FROM longnames"

df_db = pd.read_sql_query(query, conn)

conn.close()

df_db

tsn                      completename
0            50                          Bacteria
1            51                     Schizomycetes
2            52                     Archangiaceae
3            53                   Pseudomonadales
4            54                 Rhodobacteriineae
...         ...                               ...
947460  1237338  Sorghum afrorum var. bicarinatum
947461  1237339       Holcus sorghum var. afrorum
947462  1237340                     Talinum afrum
947463  1237341                          Vellassa
947464  1237342                 Vellassa maunieri

[947465 rows x 2 columns]

In [4]:
df_merged = pd.merge(df_csv, df_db, left_on = 'Taxon', right_on = 'completename')
df_merged

Year App.                     Taxon Class          Order        Family  \
0      2016    I            Aquila heliaca  Aves  Falconiformes  Accipitridae   
1      2016    I            Aquila heliaca  Aves  Falconiformes  Accipitridae   
2      2016    I      Haliaeetus albicilla  Aves  Falconiformes  Accipitridae   
3      2016    I      Haliaeetus albicilla  Aves  Falconiformes  Accipitridae   
4      2016    I      Haliaeetus albicilla  Aves  Falconiformes  Accipitridae   
...     ...  ...                       ...   ...            ...           ...   
49276  2017   II  Phaethornis atrimentalis  Aves    Apodiformes   Trochilidae   
49277  2017   II      Phaethornis hispidus  Aves    Apodiformes   Trochilidae   
49278  2017   II       Phaethornis malaris  Aves    Apodiformes   Trochilidae   
49279  2017   II         Phaethornis ruber  Aves    Apodiformes   Trochilidae   
49280  2017   II           Threnetes niger  Aves    Apodiformes   Trochilidae   

             Genus Importer Exporter Origin  Importer reported quantity  \
0           Aquila       TR       NL     CZ                         NaN   
1           Aquila       XV       RS     RS                         NaN   
2       Haliaeetus       BE       NO    NaN                         NaN   
3       Haliaeetus       BE       NO    NaN                         NaN   
4       Haliaeetus       DK       IS    NaN                       700.0   
...            ...      ...      ...    ...                         ...   
49276  Phaethornis       US       PE    NaN                         NaN   
49277  Phaethornis       US       PE    NaN                         NaN   
49278  Phaethornis       US       PE    NaN                         NaN   
49279  Phaethornis       US       PE    NaN                         NaN   
49280    Threnetes       US       PE    NaN                         NaN   

       Exporter reported quantity       Term Unit Purpose Source     tsn  \
0                             1.0     bodies  NaN       T      C  175413   
1                             1.0     bodies  NaN       Q      O  175413   
2                            43.0   feathers  NaN       S      W  175419   
3                            43.0  specimens  NaN       S      W  175419   
4                             NaN  specimens  NaN       S      W  175419   
...                           ...        ...  ...     ...    ...     ...   
49276                         3.0  specimens  NaN       S      W  693141   
49277                        13.0  specimens  NaN       S      W  555035   
49278                        43.0  specimens  NaN       S      W  555033   
49279                         3.0  specimens  NaN       S      W  555046   
49280                        16.0  specimens  NaN       S      W  555026   

                   completename  
0                Aquila heliaca  
1                Aquila heliaca  
2          Haliaeetus albicilla  
3          Haliaeetus albicilla  
4          Haliaeetus albicilla  
...                         ...  
49276  Phaethornis atrimentalis  
49277      Phaethornis hispidus  
49278       Phaethornis malaris  
49279         Phaethornis ruber  
49280           Threnetes niger  

[49281 rows x 18 columns]

In [5]:
conn = sqlite3.connect(db_path)

query = "SELECT vernacular_name, tsn FROM vernaculars WHERE language = 'English'"

df_vernaculars = pd.read_sql_query(query, conn)

conn.close()

df_vernaculars

vernacular_name      tsn
0             bacteria       50
1              diatoms     2287
2          green algae     5413
3          green algae     5414
4         brittleworts     9417
...                ...      ...
114538  lucky beantree  1237169
114539  luckybean-tree  1237169
114540       cidertree  1237327
114541          marula  1237327
114542     flameflower  1237340

[114543 rows x 2 columns]

In [6]:
df_with_human_names = pd.merge(df_merged, df_vernaculars, on = "tsn")

df_with_human_names

Year App.                 Taxon Class          Order        Family  \
0      2016    I        Aquila heliaca  Aves  Falconiformes  Accipitridae   
1      2016    I        Aquila heliaca  Aves  Falconiformes  Accipitridae   
2      2016    I        Aquila heliaca  Aves  Falconiformes  Accipitridae   
3      2016    I        Aquila heliaca  Aves  Falconiformes  Accipitridae   
4      2016    I  Haliaeetus albicilla  Aves  Falconiformes  Accipitridae   
...     ...  ...                   ...   ...            ...           ...   
71065  2017   II  Phaethornis hispidus  Aves    Apodiformes   Trochilidae   
71066  2017   II   Phaethornis malaris  Aves    Apodiformes   Trochilidae   
71067  2017   II     Phaethornis ruber  Aves    Apodiformes   Trochilidae   
71068  2017   II       Threnetes niger  Aves    Apodiformes   Trochilidae   
71069  2017   II       Threnetes niger  Aves    Apodiformes   Trochilidae   

             Genus Importer Exporter Origin  Importer reported quantity  \
0           Aquila       TR       NL     CZ                         NaN   
1           Aquila       TR       NL     CZ                         NaN   
2           Aquila       XV       RS     RS                         NaN   
3           Aquila       XV       RS     RS                         NaN   
4       Haliaeetus       BE       NO    NaN                         NaN   
...            ...      ...      ...    ...                         ...   
71065  Phaethornis       US       PE    NaN                         NaN   
71066  Phaethornis       US       PE    NaN                         NaN   
71067  Phaethornis       US       PE    NaN                         NaN   
71068    Threnetes       US       PE    NaN                         NaN   
71069    Threnetes       US       PE    NaN                         NaN   

       Exporter reported quantity       Term Unit Purpose Source     tsn  \
0                             1.0     bodies  NaN       T      C  175413   
1                             1.0     bodies  NaN       T      C  175413   
2                             1.0     bodies  NaN       Q      O  175413   
3                             1.0     bodies  NaN       Q      O  175413   
4                            43.0   feathers  NaN       S      W  175419   
...                           ...        ...  ...     ...    ...     ...   
71065                        13.0  specimens  NaN       S      W  555035   
71066                        43.0  specimens  NaN       S      W  555033   
71067                         3.0  specimens  NaN       S      W  555046   
71068                        16.0  specimens  NaN       S      W  555026   
71069                        16.0  specimens  NaN       S      W  555026   

               completename         vernacular_name  
0            Aquila heliaca    Asian Imperial Eagle  
1            Aquila heliaca  Eastern Imperial Eagle  
2            Aquila heliaca    Asian Imperial Eagle  
3            Aquila heliaca  Eastern Imperial Eagle  
4      Haliaeetus albicilla      White-tailed Eagle  
...                     ...                     ...  
71065  Phaethornis hispidus    White-bearded Hermit  
71066   Phaethornis malaris     Great-billed Hermit  
71067     Phaethornis ruber          Reddish Hermit  
71068       Threnetes niger  Pale-tailed Barbthroat  
71069       Threnetes niger        Sooty Barbthroat  

[71070 rows x 19 columns]

In [7]:
# "straight-up pandas" from here on out
df_endangered = df_with_human_names[(df_with_human_names['App.'] == "I")]

df_endangered

Year App.                 Taxon     Class          Order        Family  \
0      2016    I        Aquila heliaca      Aves  Falconiformes  Accipitridae   
1      2016    I        Aquila heliaca      Aves  Falconiformes  Accipitridae   
2      2016    I        Aquila heliaca      Aves  Falconiformes  Accipitridae   
3      2016    I        Aquila heliaca      Aves  Falconiformes  Accipitridae   
4      2016    I  Haliaeetus albicilla      Aves  Falconiformes  Accipitridae   
...     ...  ...                   ...       ...            ...           ...   
19155  2016    I     Berardius bairdii  Mammalia        Cetacea     Ziphiidae   
19156  2016    I     Berardius bairdii  Mammalia        Cetacea     Ziphiidae   
19157  2016    I              Mollusca       NaN            NaN           NaN   
19158  2016    I              Mollusca       NaN            NaN           NaN   
19159  2016    I              Reptilia  Reptilia            NaN           NaN   

            Genus Importer Exporter Origin  Importer reported quantity  \
0          Aquila       TR       NL     CZ                         NaN   
1          Aquila       TR       NL     CZ                         NaN   
2          Aquila       XV       RS     RS                         NaN   
3          Aquila       XV       RS     RS                         NaN   
4      Haliaeetus       BE       NO    NaN                         NaN   
...           ...      ...      ...    ...                         ...   
19155   Berardius       US       AU     US                        12.0   
19156   Berardius       US       AU     US                        12.0   
19157         NaN       AU       NL     XX                         NaN   
19158         NaN       AU       NL     XX                         NaN   
19159         NaN       JP       MC     XX                         2.0   

       Exporter reported quantity                      Term Unit Purpose  \
0                             1.0                    bodies  NaN       T   
1                             1.0                    bodies  NaN       T   
2                             1.0                    bodies  NaN       Q   
3                             1.0                    bodies  NaN       Q   
4                            43.0                  feathers  NaN       S   
...                           ...                       ...  ...     ...   
19155                         NaN                 specimens  NaN       S   
19156                         NaN                 specimens  NaN       S   
19157                         1.0                    shells  NaN       E   
19158                         1.0                    shells  NaN       E   
19159                         NaN  leather products (small)  NaN       Q   

      Source     tsn          completename                   vernacular_name  
0          C  175413        Aquila heliaca              Asian Imperial Eagle  
1          C  175413        Aquila heliaca            Eastern Imperial Eagle  
2          O  175413        Aquila heliaca              Asian Imperial Eagle  
3          O  175413        Aquila heliaca            Eastern Imperial Eagle  
4          W  175419  Haliaeetus albicilla                White-tailed Eagle  
...      ...     ...                   ...                               ...  
19155      W  180496     Berardius bairdii              Baird's Beaked Whale  
19156      W  180496     Berardius bairdii  north Pacific bottle-nosed whale  
19157      O   69458              Mollusca                          molluscs  
19158      O   69458              Mollusca                          mollusks  
19159      O  173747              Reptilia                          Reptiles  

[7498 rows x 19 columns]

In [8]:
print(df_endangered[df_endangered['Term'] == 'jewellery - ivory ']['vernacular_name'].unique())

['Asian Elephant' 'Asiatic Elephant' 'African Bush Elephant'
 'African elephant' 'African savannah elephant' 'Sperm Whale' 'cachalot']


In [9]:
# in-class insights
db_path = 'ITIS.sqlite'
conn = sqlite3.connect(db_path)

query2 = "SELECT tsn, geographic_value FROM geographic_div"

df_db2 = pd.read_sql_query(query2, conn)

conn.close()

df_db2 

tsn geographic_value
0          1805     East Pacific
1          2305    North America
2          2315    North America
3          2316    North America
4          2317    North America
...         ...              ...
446249  1237326           Africa
446250  1237327           Africa
446251  1237328           Africa
446252  1237329           Africa
446253  1237342    Southern Asia

[446254 rows x 2 columns]

In [10]:
df_final = pd.merge(df_merged, df_db2, on = 'tsn', how = 'left')
df_final

Year App.                     Taxon Class          Order        Family  \
0      2016    I            Aquila heliaca  Aves  Falconiformes  Accipitridae   
1      2016    I            Aquila heliaca  Aves  Falconiformes  Accipitridae   
2      2016    I      Haliaeetus albicilla  Aves  Falconiformes  Accipitridae   
3      2016    I      Haliaeetus albicilla  Aves  Falconiformes  Accipitridae   
4      2016    I      Haliaeetus albicilla  Aves  Falconiformes  Accipitridae   
...     ...  ...                       ...   ...            ...           ...   
75676  2017   II  Phaethornis atrimentalis  Aves    Apodiformes   Trochilidae   
75677  2017   II      Phaethornis hispidus  Aves    Apodiformes   Trochilidae   
75678  2017   II       Phaethornis malaris  Aves    Apodiformes   Trochilidae   
75679  2017   II         Phaethornis ruber  Aves    Apodiformes   Trochilidae   
75680  2017   II           Threnetes niger  Aves    Apodiformes   Trochilidae   

             Genus Importer Exporter Origin  Importer reported quantity  \
0           Aquila       TR       NL     CZ                         NaN   
1           Aquila       XV       RS     RS                         NaN   
2       Haliaeetus       BE       NO    NaN                         NaN   
3       Haliaeetus       BE       NO    NaN                         NaN   
4       Haliaeetus       DK       IS    NaN                       700.0   
...            ...      ...      ...    ...                         ...   
75676  Phaethornis       US       PE    NaN                         NaN   
75677  Phaethornis       US       PE    NaN                         NaN   
75678  Phaethornis       US       PE    NaN                         NaN   
75679  Phaethornis       US       PE    NaN                         NaN   
75680    Threnetes       US       PE    NaN                         NaN   

       Exporter reported quantity       Term Unit Purpose Source     tsn  \
0                             1.0     bodies  NaN       T      C  175413   
1                             1.0     bodies  NaN       Q      O  175413   
2                            43.0   feathers  NaN       S      W  175419   
3                            43.0  specimens  NaN       S      W  175419   
4                             NaN  specimens  NaN       S      W  175419   
...                           ...        ...  ...     ...    ...     ...   
75676                         3.0  specimens  NaN       S      W  693141   
75677                        13.0  specimens  NaN       S      W  555035   
75678                        43.0  specimens  NaN       S      W  555033   
75679                         3.0  specimens  NaN       S      W  555046   
75680                        16.0  specimens  NaN       S      W  555026   

                   completename geographic_value  
0                Aquila heliaca              NaN  
1                Aquila heliaca              NaN  
2          Haliaeetus albicilla    North America  
3          Haliaeetus albicilla    North America  
4          Haliaeetus albicilla    North America  
...                         ...              ...  
75676  Phaethornis atrimentalis    South America  
75677      Phaethornis hispidus    South America  
75678       Phaethornis malaris    South America  
75679         Phaethornis ruber    South America  
75680           Threnetes niger    South America  

[75681 rows x 19 columns]

In [12]:
# exploring trends by comparing importers with exporters (csv file) through the calculations of discrepencies between the two
df_with_human_names['quantity_discrepency'] = df_with_human_names['Importer reported quantity'] - df_with_human_names['Exporter reported quantity']

# creating a dataframe that includes rows where qualuty_discrepency includes values that are not empty (NaN)
discrepency_summary = df_with_human_names[df_with_human_names['quantity_discrepency'].notna()]

# to output large discrepencies
discrepency_summary.sort_values(by='quantity_discrepency', ascending = False)

Year App.                       Taxon        Class             Order  \
35919  2016   II           Galanthus elwesii          NaN          Liliales   
20296  2016   II            Acipenser baerii  Actinopteri  Acipenseriformes   
26692  2016   II  Alligator mississippiensis     Reptilia        Crocodylia   
26696  2016   II  Alligator mississippiensis     Reptilia        Crocodylia   
26695  2016   II  Alligator mississippiensis     Reptilia        Crocodylia   
...     ...  ...                         ...          ...               ...   
37994  2016   II  Cylindropuntia leptocaulis          NaN    Caryophyllales   
69124  2016  III            Cairina moschata         Aves      Anseriformes   
69810  2016  III             Cedrela odorata          NaN        Sapindales   
35920  2016   II           Galanthus elwesii          NaN          Liliales   
35921  2016   II           Galanthus elwesii          NaN          Liliales   

               Family           Genus Importer Exporter Origin  \
35919  Amaryllidaceae       Galanthus       JP       NL     TR   
20296   Acipenseridae       Acipenser       US       DE     IT   
26692   Alligatoridae       Alligator       HK       FR     US   
26696   Alligatoridae       Alligator       HK       FR     US   
26695   Alligatoridae       Alligator       HK       FR     US   
...               ...             ...      ...      ...    ...   
37994       Cactaceae  Cylindropuntia       JP       CN    NaN   
69124        Anatidae         Cairina       US       FR    NaN   
69810       Meliaceae         Cedrela       US       CI    NaN   
35920  Amaryllidaceae       Galanthus       NL       TR    NaN   
35921  Amaryllidaceae       Galanthus       NL       TR    NaN   

       Importer reported quantity  Exporter reported quantity       Term Unit  \
35919                   495000.00                2.750000e+05      roots  NaN   
20296                   214084.98                7.669898e+03    extract    g   
26692                   173819.00                6.080000e+02      skins  NaN   
26696                   173819.00                6.080000e+02      skins  NaN   
26695                   173819.00                6.080000e+02      skins  NaN   
...                           ...                         ...        ...  ...   
37994                    73732.00                4.170260e+05       live  NaN   
69124                   319453.00                1.000213e+06   medicine  NaN   
69810                      744.00                1.280698e+06  sawn wood   m3   
35920                   880764.00                2.923565e+06      roots  NaN   
35921                  1049397.00                3.818085e+06      roots  NaN   

      Purpose Source     tsn                completename  \
35919       T      W  810210           Galanthus elwesii   
20296       T      C  161066            Acipenser baerii   
26692       T      W  551771  Alligator mississippiensis   
26696       T      W  551771  Alligator mississippiensis   
26695       T      W  551771  Alligator mississippiensis   
...       ...    ...     ...                         ...   
37994       T      A  907718  Cylindropuntia leptocaulis   
69124       T      C  175246            Cairina moschata   
69810       T      A   29014             Cedrela odorata   
35920       T      A  810210           Galanthus elwesii   
35921       T      W  810210           Galanthus elwesii   

             vernacular_name  quantity_discrepency  
35919         giant snowdrop          2.200000e+05  
20296      Siberian sturgeon          2.064151e+05  
26692     American Alligator          1.732110e+05  
26696  Mississippi Alligator          1.732110e+05  
26695    Louisiana Alligator          1.732110e+05  
...                      ...                   ...  
37994          tesajo cactus         -3.432940e+05  
69124           Muscovy Duck         -6.807600e+05  
69810          Spanish cedar         -1.279954e+06  
35920         giant snowdrop         -2.0

In [15]:
print(discrepency_summary['quantity_discrepency'].describe().round(1))

count      11025.0
mean        -816.5
std        36976.8
min     -2768688.0
25%           -6.0
50%            0.0
75%            0.0
max       220000.0
Name: quantity_discrepency, dtype: float64


A lot of the originally occuring NaN values in "Importer reported quantity" and "Exporter reported quantity" columns are no longer present after adding a new column called "quality_discrepency", which dropped all the NaN values because of the second line of code. We can see that the first five rows (head of the table) shows positive values, which depicts that our third line of code worked for outputting values in descending order, indicating that the importer reported receiving more than the exporter reported sending. Then, we also see negative values for the last five rows (tail of the table), where it means that the importer reported receiving less than the exporter reported sending. To gain more insight, having the overall statistical summary of the quality_discrepency column gives us an idea on how the discrepency averages around -816.5 and returns minimum and maximum discrepancy values of -2768688.0 and 220000.0, respectively. The summary signifies that there was strong variability in reporting between importers and exporters.